crates/db_schema/src/viewer.rs

```diff
@@ -0,0 +1,96 @@
+use crate::{
+  newtypes::{LocalUserId, PersonId},
+  schema::community,
+  source::{local_user::LocalUser, site::Site},
+  CommunityVisibility,
+};
+use diesel::{dsl, query_dsl::methods::FilterDsl, ExpressionMethods};
+
+pub struct Viewer<L, S> {
+  person_id: Option<PersonId>,
+  local_user: L,
+  site_has_content_warning: S,
+}
```

Comment:
While the purpose of this struct is clear to me from its usage in the other changes files, the implementation makes little sense to me.

Why have a field for both `local_user` and `person_id` when the `local_user` will always have the `person_id` of the logged-in user, if any?

Also, why the generics? `local_user` is only ever set to an `Option<&LocalUser>`, `&LocalUser`, or `()`. Since `None` would cover the case where there is no local user, the `()` shouldn't be needed at all and the `&LocalUser` can just be handled with `Option<&LocalUser>` with `Some`. `site_has_content_warning` is only ever `bool` or `()`, so this could just default to `false` when it's not available in the constructor.

Reply:
I added comments to explain.

Without the generics, incorrect usage (e.g. calling `show_nsfw` without including `Site` in the constructor) would not fail to compile.